piptools/repositories/pypi.py

```diff
@@ -109,8 +123,15 @@ def get_dependencies(self, ireq):
 
         reqset = RequirementSet(self.build_dir,
                                 self.source_dir,
-                                download_dir=self._download_dir,
+                                # Pass no download_dir for editables, otherwise
+                                # pip uses "export" instead of "obtain"
+                                # (dropping the VCS revision info).
+                                download_dir=None if ireq.editable else self._download_dir,
                                 wheel_download_dir=self._wheel_download_dir,
+                                # Ignore installed packages: This is required
+                                # for the source_dir of VCS URLs to be created.
+                                ignore_installed=True,
                                 session=self.session)
         dependencies = reqset._prepare_file(self.finder, ireq)
-        return set(dependencies)
+        self._get_dependencies_cache[ireq] = set(dependencies)
+        return self._get_dependencies_cache[ireq]
```

Comment:
Here, too. Please avoid caching things in the middle of the core logic. This will quickly explode the complexity. Untangling this stuff was my main driver behind the rewrite in the first place.